packages/@datapull/chef-origin/index.js

```diff
@@ -1,36 +1,59 @@
 const util = require('util');
-const objectPath = require('object-path');
+const fs = require('fs');
 const chef_api = require("chef-api");
+const Bottleneck = require("bottleneck");
 
 class ChefOrigin {
 
-    constructor(config) {
-        this.config = config;
-    }
+    constructor(stepConfig) {}
 
     get originDeclaration() {
         return {
-            name: ['Chef', this.config.organization].join(':'),
+            name: ['Chef'].join(':'),
             runner: this.pullData.bind(this)
         };
     }
 
-    async pullData(config) {
-        console.log('[Chef] pulling data for ', this.config.organization);
+    async pullData(pipelineConfig) {
+        console.log('[Chef] pulling data for ', pipelineConfig.apiUrl);
+
+        if (!pipelineConfig.key_path) {
+          console.error(`[Chef origin] key path is not specified`);
+          throw Error(`Key path is not specified`);
+        }
+
+        if (!fs.existsSync(pipelineConfig.key_path)) {
+          console.error(`[Chef origin] key does not exist in the given path ${pipelineConfig.key_path}`);
+          throw Error(`Key does not exist in the given path ${pipelineConfig.key_path}`);
+        }
 
         const options = {
-            user_name: config.username,
-            key_path: config.key_path,
-            url: config.apiUrl
+            user_name: pipelineConfig.username,
+            key_path: pipelineConfig.key_path,
+            url: pipelineConfig.apiUrl
         };
 
         const chef = new chef_api();
         chef.config(options);
-        const nodes = await util.promisify(chef.getNodes)()
+        const nodes = await util.promisify(chef.getNodes)();
         const results = [];
+
+        // setup rate limiter:
+        const rateLimitConfig = pipelineConfig.rateLimiter || {};
+        const maxConcurrent = rateLimitConfig.maxConcurrent || 10;
+        const minTime = rateLimitConfig.minTimeBetweenRequestsMs || 200;
+        const highWater = rateLimitConfig.highWater || -1;
+        const rateLimitStrategy = rateLimitConfig.rateLimitStrategy || 'BLOCK';
+        const rejectOnDrop = 'rejectOnDrop' in rateLimitConfig ? rateLimitConfig.rejectOnDrop : true;
+
+        const limiter = new Bottleneck(maxConcurrent, minTime, highWater, Bottleneck.strategy[rateLimitStrategy], rejectOnDrop);
+
+        // queue all the nodes:
         for (const nodeName in nodes) {
             if (nodes.hasOwnProperty(nodeName)) {
-                const node = util.promisify(chef.getNode)(nodeName);
+                const node = limiter.schedule(function () {
+                  return util.promisify(chef.getNode)(nodeName);
```

Comment:
I think it's better not to wrap the call in `promisify` and then wrap it up in function, but rather just use `limiter.schedule(chef.getNode, nodeName)`.